### DISCUSSION TITLE
Realtime Velocity and Position control

### DISCUSSION THREAD

**Discussion by kalepo** [ros-industrial/abb_robot_driver#72] [Q&A]
Im trying to achieve real time control of my robot arm in ROS. 
Im currently able to send positions into the position controller but its very slow. 
I want to be able to achieve real time control. 
How would I go about to do this?

**Comment by gavanderhoorn** [comment]
Please start by describing in detail what you've done so far, and what result that had.

Include information about used versions, configurations of controllers (both ROS and the IRC), etc.

And please clarify how this is related to #63.

**Reply by kalepo** [reply]
I have currently been able to subscribe to a sensor_msg/JointState message topic from Unity and then published the positions and velocities of the URDF robot into the topic /egm/joint_group_velocity_controller/command using message trajectory_msgs/JointTrajectory. 

When I run the simulation, the message comes up with Unexpected Error: No trajectory defined at current time. Please contact the package maintainer. 

And then the robot in robotstudio has a system failure. 

Any ideas on what is going wrong?

Ros is Ros Melodic and this is using the StateMachine 1.1 in RobotStudio on an IRB1200

**Reply by gavanderhoorn** [reply]
> I have currently been able to subscribe to a sensor_msg/JointState message topic from Unity and then published the positions and velocities of the URDF robot into the topic /egm/joint_group_velocity_controller/command using message trajectory_msgs/JointTrajectory.

this confuses me.

Unless you've changed something, I'd expect `/egm/joint_group_velocity_controller/command` to accept velocity commands *only*, not "positions and velocities". It's a *velocity controller* after all.

Please describe *how* exactly you are converting the data.

Also: `trajectory_msgs/JointTrajectory` is not a message you can publish to `/egm/joint_group_velocity_controller/command`, unless `joint_group_velocity_controller` is actually a `JointTrajectoryController`.

So please verify your description is correct, perhaps by showing the output of `rqt_graph` and/or some suitable `rostopic` and `rosnode` commands.

> When I run the simulation, the message comes up with Unexpected Error: No trajectory defined at current time. Please contact the package maintainer.

As your description has me confused (as in: it should not be possible to do what you describe), I don't believe it would help to go into specific error messages such as this one.

---

Having written that: if your goal is to replicate motion / the state of your robot in Unity, position control would probably be a better choice, as at least those are absolute set points (instead of some relative velocity command which would have to be integrated to end up with position).

I would:

 1. configure `abb_robot_driver` to load a `joint_group_position_controller`
 2. write a simple node that subscribes to the `JointState` messages published by your Unity simulation and converts the data in the `positions` field to a `std_msgs/Float64MultiArray` (make sure to use the correct joint order)
 3. publish that message to the `command` topic of the `joint_group_position_controller`

*If* the data coming out of your Unity simulation is smooth and consistent, that *should* make your RobotStudio simulation follow whatever happens in Unity.

Be sure to update the settings of your RobotStudio simulated controller though, as by default things like joint limits are set to very low values.

**Reply by kalepo** [reply]
Apologies, I just changed the controller to a joint trajectory controller but left the name of it. 

I tried the position controller, but it is quite slow. How would i increase the speed of the movements?

**Reply by gavanderhoorn** [reply]
As I wrote before: please provide more information.

You are not answering my questions, nor do you describe what you've done.

Without more information your question cannot be answered.

**Reply by kalepo** [reply]
Apologies. Let me try to explain systematically

Method 1 - publishing position values to position controller

1. Published joint states of URDF from unity into a topic x through sensor_msgs/JointState
2. Loaded joint_group_position_controller
3. Subscribed to that topic x and published position values to the topic /egm/joint_group_position_controller/command through std_msgs/Float64MultiArray
4. The positions are able to be moved from unity to robotstudio but they are moving very slowly. How do I speed this up?

Method 2- publishing positions and velocities to trajectory controller 
1. Published joint states of URDF from unity into a topic x. This uses sensor_msgs/JointState.
2. Loaded position_velocity_controller
3. Subscribed to that topic x and published position and velocity values to the topic /egm/position_velocity_controller/command through trajectory_msgs/jointtrajectory.
4. Then when I move the positions in unity, there is an 
Unexpected Error: No trajectory defined at current time. Please contact the package maintainer.
5. The robot in robot studio does not move when I change the position of a joint in unity

**Reply by gavanderhoorn** [reply]
Using anything but a `JointGroupPositionController` for your use-case makes little sense to me -- from the standard set of controllers offered by `ros_control` anyway.

Have you checked the [Troubleshooting](https://github.com/ros-industrial/abb_robot_driver/tree/32eb0f48abfd3358cb5a4f3f043f1306a9811b24/abb_egm_hardware_interface#troubleshooting) section in the readme of `abb_egm_hardware_interface`? Specifically the section about motion? If you haven't changed the defaults for `MaxSpeedDeviation` for instance, the robot will move very slowly indeed.

(I've responded here to keep the thread in one place)

**Reply by kalepo** [reply]
Yup I have checked the troubleshooting, where would I find the MaxSpeedDeviation? Is it within the ROS Side or on the robotstudio side?

**Reply by chastjj** [reply]
you can modify the default in the rapid settings
![{BFE47E3E-41A6-4df3-B4FE-AD261DA077BC}](https://github.com/user-attachments/assets/62858a61-b377-4b73-bea8-15e679a083a4)

**Reply by gavanderhoorn** [reply]
I would probably recommend using the `get_egm_settings` and `set_egm_settings` services instead.

**Reply by chastjj** [reply]
It was strange that when I use set_egm_settings to set the max_speed_deviation, the robot speed seems no change, I try to set speed ratio by /rws/set_speed_ratio,  it does not change speed of robot, I use position trajectory controller in moveit,  but it seems the speed only affect by the settings I modified in  robot studio rapid settings.

**Reply by gavanderhoorn** [reply]
Changing the settings can only be done *before* starting a new EGM session.

> I try to set speed ratio by /rws/set_speed_ratio, it does not change speed of robot

AFAIK (and IIRC), final EGM velocity is not influenced by the speed override (or robot speed). You must change the EGM settings for that.

**Reply by gavanderhoorn** [reply]
@Chastj: you may be interested in [sam-xl/abb_robot_driver_state_machine](https://github.com/sam-xl/abb_robot_driver_state_machine). It automates some of the steps needed to use `abb_robot_driver`.

It's ROS 1 only though.

**Reply by chastjj** [reply]
> Changing the settings can only be done _before_ starting a new EGM session.
> 
> > I try to set speed ratio by /rws/set_speed_ratio, it does not change speed of robot
> 
> AFAIK (and IIRC), final EGM velocity is not influenced by the speed override (or robot speed). You must change the EGM settings for that.

I find that change the `max_speed_deviation` in robot studio rapid settings can affect robot speed when use `position_trajectory controller `in moveit, but when use `set_egm_settings` to change the `max_speed_deviation`, it can't change the robot speed. what can I do if I want to change robot speed when use moveit?

**Reply by gavanderhoorn** [reply]
I would suggest to take a look at the order of operations (ie: service calls) the already mentioned [abb_robot_driver_state_machine](https://github.com/sam-xl/abb_robot_driver_state_machine) uses.

It is capable of changing the `max_speed_deviation` using the `set_egm_settings` service.

**Comment by kalepo** [comment]
Hi @gavanderhoorn whats your thoughts on this?
